okio/src/jvmMain/kotlin/okio/Pipe.kt

```diff
@@ -181,26 +182,30 @@ class Pipe(internal val maxBufferSize: Long) {
           throw IOException("canceled")
         }
 
+        closed = sinkClosed
         if (buffer.exhausted()) {
           sourceClosed = true
           foldedSink = sink
-          return@fold
+          done = true
+          return@withLock
         }
 
-        closed = sinkClosed
         sinkBuffer = Buffer()
         sinkBuffer.write(buffer, buffer.size)
         condition.signalAll() // Notify the sink that it can resume writing.
       }
 
-      var success = false
-      try {
-        sink.write(sinkBuffer, sinkBuffer.size)
+      if (done) {
         if (closed) {
           sink.close()
-        } else {
-          sink.flush()
```

Comment:
It wasn't clear if this flush vs. close was really necessary. It felt like an optimization. I changed it because I didn't want the code to inadvertently close the sink 2x, and also didn't want to add yet-another piece of state to prevent such a double-close. The code seemed a little easier on the eyes to simply have the close called in one place, and accept the potential overhead and explicit flush right before close.

Reply:
> inadvertently close the sink 2x

The [documentation of `Sink`](https://square.github.io/okio/3.x/okio/okio/okio/-sink/close.html) requires the second call to be a no-op. So if it simplifies things here, you are free to do it.

Reply:
I don't know that it simplifies anything to let it call `close` here, too. It would be an extra conditional and a couple of extra lines that IMO aren't much value. Feels like we should only be bothering to call `close` it at the end.

I was just explaining that by moving the call to close to the end, we _could_ end up calling flush needlessly, because we'll call it in the last iteration of the loop for the last chunk of data, but we might then immediately call close on return. We don't know for certain here whether we're going to close, so we do need the flush. A possible alternative would be to defer calling flush until we return from the loop, but then we might induce unsynchronized concurrent access to the destination sink -- by calling flush in the folding thread, which is stated to be safe to call concurrently with calls to write to the pipe's sink, which would also write to the destination sink, but from a different thread.

Sorry if that's hard to parse. Maybe it would be more clear to just review the change and see if it looks fine or if you'd prefer an alternate formulation. (It's small; most of the lines of code in this PR are the new test case.)

Reply:
I'm on mobile which is why I didn't. I'll reread this and review the diff later or tomorrow on a nice big screen.

Reply:
(gentle reminder, if you have a moment to look)

Reply:
I did take a look, and it only served to convince me that @swankjesse should be the one looking!